// tgsnake/lib/src/TL/Messages/Medias/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LinkPreviewOptions = exports.WebPage = exports.Voice = exports.VideoNote = exports.Video = exports.Location = exports.Venue = exports.Sticker = exports.PollOption = exports.PollAnswer = exports.Poll = exports.PhotoSize = exports.Photo = exports.Game = exports.Document = exports.Dice = exports.Contact = exports.Audio = exports.Animation = void 0;
var Animated_js_1 = require("./Animated.js");
Object.defineProperty(exports, "Animation", { enumerable: true, get: function () { return Animated_js_1.Animation; } });
var Audio_js_1 = require("./Audio.js");
Object.defineProperty(exports, "Audio", { enumerable: true, get: function () { return Audio_js_1.Audio; } });
var Contact_js_1 = require("./Contact.js");
Object.defineProperty(exports, "Contact", { enumerable: true, get: function () { return Contact_js_1.Contact; } });
var Dice_js_1 = require("./Dice.js");
Object.defineProperty(exports, "Dice", { enumerable: true, get: function () { return Dice_js_1.Dice; } });
var Document_js_1 = require("./Document.js");
Object.defineProperty(exports, "Document", { enumerable: true, get: function () { return Document_js_1.Document; } });
var Game_js_1 = require("./Game.js");
Object.defineProperty(exports, "Game", { enumerable: true, get: function () { return Game_js_1.Game; } });
var Photo_js_1 = require("./Photo.js");
Object.defineProperty(exports, "Photo", { enumerable: true, get: function () { return Photo_js_1.Photo; } });
var PhotoSize_js_1 = require("./PhotoSize.js");
Object.defineProperty(exports, "PhotoSize", { enumerable: true, get: function () { return PhotoSize_js_1.PhotoSize; } });
var Poll_js_1 = require("./Poll.js");
Object.defineProperty(exports, "Poll", { enumerable: true, get: function () { return Poll_js_1.Poll; } });
Object.defineProperty(exports, "PollAnswer", { enumerable: true, get: function () { return Poll_js_1.PollAnswer; } });
Object.defineProperty(exports, "PollOption", { enumerable: true, get: function () { return Poll_js_1.PollOption; } });
var Sticker_js_1 = require("./Sticker.js");
Object.defineProperty(exports, "Sticker", { enumerable: true, get: function () { return Sticker_js_1.Sticker; } });
var Venue_js_1 = require("./Venue.js");
Object.defineProperty(exports, "Venue", { enumerable: true, get: function () { return Venue_js_1.Venue; } });
Object.defineProperty(exports, "Location", { enumerable: true, get: function () { return Venue_js_1.Location; } });
var Video_js_1 = require("./Video.js");
Object.defineProperty(exports, "Video", { enumerable: true, get: function () { return Video_js_1.Video; } });
var VideoNote_js_1 = require("./VideoNote.js");
Object.defineProperty(exports, "VideoNote", { enumerable: true, get: function () { return VideoNote_js_1.VideoNote; } });
var Voice_js_1 = require("./Voice.js");
Object.defineProperty(exports, "Voice", { enumerable: true, get: function () { return Voice_js_1.Voice; } });
var WebPage_js_1 = require("./WebPage.js");
Object.defineProperty(exports, "WebPage", { enumerable: true, get: function () { return WebPage_js_1.WebPage; } });
var LinkPreviewOptions_js_1 = require("./LinkPreviewOptions.js");
Object.defineProperty(exports, "LinkPreviewOptions", { enumerable: true, get: function () { return LinkPreviewOptions_js_1.LinkPreviewOptions; } });
