// tgsnake/lib/src/TL/Messages/Medias/LinkPreviewOptions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LinkPreviewOptions = void 0;
const platform_node_js_1 = require("../../../platform.node.js");
const TL_js_1 = require("../../TL.js");
class LinkPreviewOptions extends TL_js_1.TLObject {
    url;
    preferSmallMedia;
    preferLargeMedia;
    showAboveText;
    constructor({ url, preferSmallMedia, preferLargeMedia, showAboveText, }, client) {
        super(client);
        this.url = url;
        this.preferSmallMedia = preferSmallMedia;
        this.preferLargeMedia = preferLargeMedia;
        this.showAboveText = showAboveText;
    }
    static parse(client, webpage) {
        const parsed = new LinkPreviewOptions({
            preferLargeMedia: webpage.forceLargeMedia,
            preferSmallMedia: webpage.forceSmallMedia,
        }, client);
        if (webpage.webpage instanceof platform_node_js_1.Raw.WebPageEmpty) {
            parsed.url = webpage.webpage.url;
        }
        else if (webpage.webpage instanceof platform_node_js_1.Raw.WebPagePending) {
            parsed.url = webpage.webpage.url;
        }
        else if (webpage.webpage instanceof platform_node_js_1.Raw.WebPage) {
            parsed.url = webpage.webpage.url;
        }
        return parsed;
    }
}
exports.LinkPreviewOptions = LinkPreviewOptions;
