// @tgsnake/core/lib/src/errors/SecretChat.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AlreadyAccepted = exports.ChatNotFound = exports.FingerprintMismatch = exports.SecretChatError = void 0;
const platform_node_js_1 = require("../platform.node.js");
class SecretChatError extends Error {
    message;
    description;
    constructor(message, description) {
        super();
        this.message = message;
        this.description = description;
    }
    [Symbol.for('nodejs.util.inspect.custom')]() {
        const toPrint = {
            _: this.constructor.name,
        };
        for (const key in this) {
            if (this.hasOwnProperty(key)) {
                const value = this[key];
                if (!key.startsWith('_') && value !== undefined && value !== null) {
                    toPrint[key] = value;
                }
            }
        }
        Object.setPrototypeOf(toPrint, {
            stack: this.stack,
        });
        return toPrint;
    }
    [Symbol.for('Deno.customInspect')]() {
        return String((0, platform_node_js_1.inspect)(this[Symbol.for('nodejs.util.inspect.custom')](), { colors: true }));
    }
    toJSON() {
        const toPrint = {
            _: this.constructor.name,
            stack: this.stack,
        };
        for (const key in this) {
            if (this.hasOwnProperty(key)) {
                const value = this[key];
                if (!key.startsWith('_') && value !== undefined && value !== null) {
                    toPrint[key] = typeof value === 'bigint' ? String(value) : value;
                }
            }
        }
        return toPrint;
    }
    toString() {
        return `[constructor of ${this.constructor.name}] ${JSON.stringify(this, null, 2)}`;
    }
}
exports.SecretChatError = SecretChatError;
class FingerprintMismatch extends SecretChatError {
    constructor() {
        super('Fingerprint key mismatch', 'Given fingerprint key from message is mismatch. So the message is not secure and the secret chat should be closed.');
    }
}
exports.FingerprintMismatch = FingerprintMismatch;
class ChatNotFound extends SecretChatError {
    constructor(chatId) {
        super('Secret chat not found', `Provided chatId (${chatId}) is not found in session. Make sure the chatId is correct and already saved in session.`);
    }
}
exports.ChatNotFound = ChatNotFound;
class AlreadyAccepted extends SecretChatError {
    constructor() {
        super('Secret chat already accepted');
    }
}
exports.AlreadyAccepted = AlreadyAccepted;
