// @tgsnake/core/lib/src/errors/WebSocket.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProxyUnsupported = exports.ReadClosed = exports.Disconnected = exports.WebSocketError = void 0;
const platform_node_js_1 = require("../platform.node.js");
class WebSocketError extends Error {
    message;
    description;
    constructor(message, description) {
        super();
        this.message = message;
        this.description = description;
    }
    [Symbol.for('nodejs.util.inspect.custom')]() {
        const toPrint = {
            _: this.constructor.name,
        };
        for (const key in this) {
            if (this.hasOwnProperty(key)) {
                const value = this[key];
                if (!key.startsWith('_') && value !== undefined && value !== null) {
                    toPrint[key] = value;
                }
            }
        }
        Object.setPrototypeOf(toPrint, {
            stack: this.stack,
        });
        return toPrint;
    }
    [Symbol.for('Deno.customInspect')]() {
        return String((0, platform_node_js_1.inspect)(this[Symbol.for('nodejs.util.inspect.custom')](), { colors: true }));
    }
    toJSON() {
        const toPrint = {
            _: this.constructor.name,
            stack: this.stack,
        };
        for (const key in this) {
            if (this.hasOwnProperty(key)) {
                const value = this[key];
                if (!key.startsWith('_') && value !== undefined && value !== null) {
                    toPrint[key] = typeof value === 'bigint' ? String(value) : value;
                }
            }
        }
        return toPrint;
    }
    toString() {
        return `[constructor of ${this.constructor.name}] ${JSON.stringify(this, null, 2)}`;
    }
}
exports.WebSocketError = WebSocketError;
class Disconnected extends WebSocketError {
    constructor() {
        super('WebSocket Disconnected', "This happen when you trying to send request or receive update from websocket server hut the websocket client is doesn't ready. Make sure the websocket client is connected to server.");
    }
}
exports.Disconnected = Disconnected;
class ReadClosed extends WebSocketError {
    constructor() {
        super('WebSocket connection closed when reading data', 'This happen when suddenly the connection between the websocket client and the server is lost when fetching data updates from the server.');
    }
}
exports.ReadClosed = ReadClosed;
class ProxyUnsupported extends WebSocketError {
    constructor() {
        super('WebSocket proxy unsupported', 'This is because browser telegram or websocket proxy are not supported by the framework at this time.');
    }
}
exports.ProxyUnsupported = ProxyUnsupported;
