// tgsnake/lib/src/TL/Advanced/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WebAppData = exports.User = exports.Restriction = exports.ChatPhoto = exports.ChatPermission = exports.Chat = void 0;
var Chat_js_1 = require("./Chat.js");
Object.defineProperty(exports, "Chat", { enumerable: true, get: function () { return Chat_js_1.Chat; } });
var ChatPermission_js_1 = require("./ChatPermission.js");
Object.defineProperty(exports, "ChatPermission", { enumerable: true, get: function () { return ChatPermission_js_1.ChatPermission; } });
var ChatPhoto_js_1 = require("./ChatPhoto.js");
Object.defineProperty(exports, "ChatPhoto", { enumerable: true, get: function () { return ChatPhoto_js_1.ChatPhoto; } });
var Restriction_js_1 = require("./Restriction.js");
Object.defineProperty(exports, "Restriction", { enumerable: true, get: function () { return Restriction_js_1.Restriction; } });
var User_js_1 = require("./User.js");
Object.defineProperty(exports, "User", { enumerable: true, get: function () { return User_js_1.User; } });
var WebApp_js_1 = require("./WebApp.js");
Object.defineProperty(exports, "WebAppData", { enumerable: true, get: function () { return WebApp_js_1.WebAppData; } });
