// @tgsnake/core/lib/src/errors/exceptions/SeeOther303.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UserMigrate = exports.StatsMigrate = exports.PhoneMigrate = exports.NetworkMigrate = exports.FileMigrate = exports.SeeOther = void 0;
const RpcError_js_1 = require("../RpcError.js");
class SeeOther extends RpcError_js_1.RPCError {
    code = 303;
    name = 'SEE_OTHER';
}
exports.SeeOther = SeeOther;
class FileMigrate extends SeeOther {
    id = 'FILE_MIGRATE_X';
    message = 'The file to be accessed is currently stored in DC{value}';
}
exports.FileMigrate = FileMigrate;
class NetworkMigrate extends SeeOther {
    id = 'NETWORK_MIGRATE_X';
    message = 'The source IP address is associated with DC{value} (for registration)';
}
exports.NetworkMigrate = NetworkMigrate;
class PhoneMigrate extends SeeOther {
    id = 'PHONE_MIGRATE_X';
    message = 'The phone number a user is trying to use for authorization is associated with DC{value}';
}
exports.PhoneMigrate = PhoneMigrate;
class StatsMigrate extends SeeOther {
    id = 'STATS_MIGRATE_X';
    message = 'The statistics of the group/channel are stored in DC{value}';
}
exports.StatsMigrate = StatsMigrate;
class UserMigrate extends SeeOther {
    id = 'USER_MIGRATE_X';
    message = 'The user whose identity is being used to execute queries is associated with DC{value} (for registration)';
}
exports.UserMigrate = UserMigrate;
