// tgsnake/lib/src/Methods/Messages/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.sendPhoto = exports.deleteMessages = exports.download = exports.sendAudio = exports.sendSticker = exports.sendAnimation = exports.sendVideoNote = exports.sendVideo = exports.sendDocument = exports.sendMedia = exports.sendMessage = exports.getMessages = void 0;
var GetMessages_js_1 = require("./GetMessages.js");
Object.defineProperty(exports, "getMessages", { enumerable: true, get: function () { return GetMessages_js_1.getMessages; } });
var SendMessage_js_1 = require("./SendMessage.js");
Object.defineProperty(exports, "sendMessage", { enumerable: true, get: function () { return SendMessage_js_1.sendMessage; } });
var SendMedia_js_1 = require("./SendMedia.js");
Object.defineProperty(exports, "sendMedia", { enumerable: true, get: function () { return SendMedia_js_1.sendMedia; } });
var SendDocument_js_1 = require("./SendDocument.js");
Object.defineProperty(exports, "sendDocument", { enumerable: true, get: function () { return SendDocument_js_1.sendDocument; } });
var SendVideo_js_1 = require("./SendVideo.js");
Object.defineProperty(exports, "sendVideo", { enumerable: true, get: function () { return SendVideo_js_1.sendVideo; } });
var SendVideoNote_js_1 = require("./SendVideoNote.js");
Object.defineProperty(exports, "sendVideoNote", { enumerable: true, get: function () { return SendVideoNote_js_1.sendVideoNote; } });
var SendAnimation_js_1 = require("./SendAnimation.js");
Object.defineProperty(exports, "sendAnimation", { enumerable: true, get: function () { return SendAnimation_js_1.sendAnimation; } });
var SendSticker_js_1 = require("./SendSticker.js");
Object.defineProperty(exports, "sendSticker", { enumerable: true, get: function () { return SendSticker_js_1.sendSticker; } });
var SendAudio_js_1 = require("./SendAudio.js");
Object.defineProperty(exports, "sendAudio", { enumerable: true, get: function () { return SendAudio_js_1.sendAudio; } });
var Download_js_1 = require("./Download.js");
Object.defineProperty(exports, "download", { enumerable: true, get: function () { return Download_js_1.download; } });
var DeleteMessage_js_1 = require("./DeleteMessage.js");
Object.defineProperty(exports, "deleteMessages", { enumerable: true, get: function () { return DeleteMessage_js_1.deleteMessages; } });
var SendPhoto_js_1 = require("./SendPhoto.js");
Object.defineProperty(exports, "sendPhoto", { enumerable: true, get: function () { return SendPhoto_js_1.sendPhoto; } });
