// tgsnake/lib/src/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Plugins = exports.Versions = exports.Utilities = exports.Methods = exports.TLs = exports.Conversations = exports.Contexts = exports.Clients = exports.Telegram = exports.Raw = exports.Composer = exports.SnakeSession = exports.shutdown = exports.Snake = void 0;
var Snake_js_1 = require("./Client/Snake.js");
Object.defineProperty(exports, "Snake", { enumerable: true, get: function () { return Snake_js_1.Snake; } });
Object.defineProperty(exports, "shutdown", { enumerable: true, get: function () { return Snake_js_1.shutdown; } });
var SnakeSession_js_1 = require("./Client/SnakeSession.js");
Object.defineProperty(exports, "SnakeSession", { enumerable: true, get: function () { return SnakeSession_js_1.SnakeSession; } });
var Composer_js_1 = require("./Context/Composer.js");
Object.defineProperty(exports, "Composer", { enumerable: true, get: function () { return Composer_js_1.Composer; } });
var platform_node_js_1 = require("./platform.node.js");
Object.defineProperty(exports, "Raw", { enumerable: true, get: function () { return platform_node_js_1.Raw; } });
var Telegram_js_1 = require("./Methods/Telegram.js");
Object.defineProperty(exports, "Telegram", { enumerable: true, get: function () { return Telegram_js_1.Telegram; } });
exports.Clients = __importStar(require("./Client/index.js"));
exports.Contexts = __importStar(require("./Context/index.js"));
exports.Conversations = __importStar(require("./Conversation/index.js"));
exports.TLs = __importStar(require("./TL/index.js"));
exports.Methods = __importStar(require("./Methods/index.js"));
exports.Utilities = __importStar(require("./Utilities.js"));
exports.Versions = __importStar(require("./Version.node.js"));
exports.Plugins = __importStar(require("./Plugins/index.js"));
