// tgsnake/lib/src/platform.node.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.sysprc = exports.isDeno = exports.isBrowser = exports.cwd = exports.fs = exports.prompts = exports.path = exports.Buffer = exports.Duplex = exports.Writable = exports.Readable = exports.mimetypes = exports.Logger = exports.Parser = exports.DOCUMENT_TYPES = exports.PHOTO_TYPES = exports.FileTypeUniqueId = exports.ThumbnailSource = exports.FileType = exports.FileId = exports.base64_url_encode = exports.Writer = exports.Errors = exports.Cryptos = exports.Files = exports.Client = exports.Sessions = exports.Clients = exports.Raws = exports.Storages = exports.Helpers = exports.Raw = void 0;
const path = __importStar(require("path"));
exports.path = path;
const fs = __importStar(require("fs"));
exports.fs = fs;
const prompts_1 = __importDefault(require("prompts"));
exports.prompts = prompts_1.default;
var core_1 = require("@tgsnake/core");
Object.defineProperty(exports, "Raw", { enumerable: true, get: function () { return core_1.Raw; } });
Object.defineProperty(exports, "Helpers", { enumerable: true, get: function () { return core_1.Helpers; } });
Object.defineProperty(exports, "Storages", { enumerable: true, get: function () { return core_1.Storages; } });
Object.defineProperty(exports, "Raws", { enumerable: true, get: function () { return core_1.Raws; } });
Object.defineProperty(exports, "Clients", { enumerable: true, get: function () { return core_1.Clients; } });
Object.defineProperty(exports, "Sessions", { enumerable: true, get: function () { return core_1.Sessions; } });
Object.defineProperty(exports, "Client", { enumerable: true, get: function () { return core_1.Client; } });
Object.defineProperty(exports, "Files", { enumerable: true, get: function () { return core_1.Files; } });
Object.defineProperty(exports, "Cryptos", { enumerable: true, get: function () { return core_1.Cryptos; } });
Object.defineProperty(exports, "Errors", { enumerable: true, get: function () { return core_1.Errors; } });
var fileid_1 = require("@tgsnake/fileid");
Object.defineProperty(exports, "Writer", { enumerable: true, get: function () { return fileid_1.Writer; } });
Object.defineProperty(exports, "base64_url_encode", { enumerable: true, get: function () { return fileid_1.base64_url_encode; } });
Object.defineProperty(exports, "FileId", { enumerable: true, get: function () { return fileid_1.FileId; } });
Object.defineProperty(exports, "FileType", { enumerable: true, get: function () { return fileid_1.FileType; } });
Object.defineProperty(exports, "ThumbnailSource", { enumerable: true, get: function () { return fileid_1.ThumbnailSource; } });
Object.defineProperty(exports, "FileTypeUniqueId", { enumerable: true, get: function () { return fileid_1.FileTypeUniqueId; } });
Object.defineProperty(exports, "PHOTO_TYPES", { enumerable: true, get: function () { return fileid_1.PHOTO_TYPES; } });
Object.defineProperty(exports, "DOCUMENT_TYPES", { enumerable: true, get: function () { return fileid_1.DOCUMENT_TYPES; } });
var parser_1 = require("@tgsnake/parser");
Object.defineProperty(exports, "Parser", { enumerable: true, get: function () { return parser_1.Parser; } });
var log_1 = require("@tgsnake/log");
Object.defineProperty(exports, "Logger", { enumerable: true, get: function () { return log_1.Logger; } });
var mime_types_1 = require("mime-types");
Object.defineProperty(exports, "mimetypes", { enumerable: true, get: function () { return mime_types_1.lookup; } });
var stream_1 = require("stream");
Object.defineProperty(exports, "Readable", { enumerable: true, get: function () { return stream_1.Readable; } });
Object.defineProperty(exports, "Writable", { enumerable: true, get: function () { return stream_1.Writable; } });
Object.defineProperty(exports, "Duplex", { enumerable: true, get: function () { return stream_1.Duplex; } });
var buffer_1 = require("buffer");
Object.defineProperty(exports, "Buffer", { enumerable: true, get: function () { return buffer_1.Buffer; } });
exports.cwd = process.cwd;
exports.isBrowser = typeof window !== 'undefined';
exports.isDeno = false;
exports.sysprc = process;
