// @tgsnake/fileid/lib/index.js
"use strict";
/**
 * tgsnake - Telegram MTProto framework for nodejs.
 * Copyright (C) 2022 butthx <https://github.com/butthx>
 *
 * THIS FILE IS PART OF TGSNAKE
 *
 * tgsnake is a free software : you can redistribute it and/or modify
 * it under the terms of the MIT License as published.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Decode = exports.Encode = exports.FileId = exports.DOCUMENT_TYPES = exports.PHOTO_TYPES = exports.Reader = exports.Writer = exports.rle_decode = exports.rle_encode = exports.base64_url_decode = exports.base64_url_encode = exports.FileTypeUniqueId = exports.ThumbnailSource = exports.FileType = void 0;
var utils_js_1 = require("./utils.js");
Object.defineProperty(exports, "FileType", { enumerable: true, get: function () { return utils_js_1.FileType; } });
Object.defineProperty(exports, "ThumbnailSource", { enumerable: true, get: function () { return utils_js_1.ThumbnailSource; } });
Object.defineProperty(exports, "FileTypeUniqueId", { enumerable: true, get: function () { return utils_js_1.FileTypeUniqueId; } });
Object.defineProperty(exports, "base64_url_encode", { enumerable: true, get: function () { return utils_js_1.base64_url_encode; } });
Object.defineProperty(exports, "base64_url_decode", { enumerable: true, get: function () { return utils_js_1.base64_url_decode; } });
Object.defineProperty(exports, "rle_encode", { enumerable: true, get: function () { return utils_js_1.rle_encode; } });
Object.defineProperty(exports, "rle_decode", { enumerable: true, get: function () { return utils_js_1.rle_decode; } });
Object.defineProperty(exports, "Writer", { enumerable: true, get: function () { return utils_js_1.Writer; } });
Object.defineProperty(exports, "Reader", { enumerable: true, get: function () { return utils_js_1.Reader; } });
Object.defineProperty(exports, "PHOTO_TYPES", { enumerable: true, get: function () { return utils_js_1.PHOTO_TYPES; } });
Object.defineProperty(exports, "DOCUMENT_TYPES", { enumerable: true, get: function () { return utils_js_1.DOCUMENT_TYPES; } });
var parser_js_1 = require("./parser.js");
Object.defineProperty(exports, "FileId", { enumerable: true, get: function () { return parser_js_1.FileId; } });
var encode_js_1 = require("./encode.js");
Object.defineProperty(exports, "Encode", { enumerable: true, get: function () { return encode_js_1.Encode; } });
var decode_js_1 = require("./decode.js");
Object.defineProperty(exports, "Decode", { enumerable: true, get: function () { return decode_js_1.Decode; } });
