// @tgsnake/core/lib/src/raw/core/Message.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Message = void 0;
const BytesIO_js_1 = require("./BytesIO.js");
const TLObject_js_1 = require("./TLObject.js");
const Primitive = __importStar(require("./primitive/index.js"));
const platform_node_js_1 = require("../../platform.node.js");
function toBytes(value) {
    const bytesArray = [];
    for (let i = 0; i < 8; i++) {
        let shift = value >> BigInt(8 * i);
        shift &= BigInt(255);
        bytesArray[i] = Number(String(shift));
    }
    return platform_node_js_1.Buffer.from(bytesArray);
}
class Message extends TLObject_js_1.TLObject {
    static ID = 0x5bb8e511;
    msgId;
    seqNo;
    length;
    body;
    constructor(body, msgId, seqNo, length) {
        super();
        this.className = 'Message';
        this._slots = ['body', 'msgId', 'seqNo', 'length'];
        this.msgId = msgId;
        this.seqNo = seqNo;
        this.length = length;
        this.body = body;
    }
    static async read(data, ..._args) {
        let msgId = await Primitive.Long.read(data);
        let seqNo = await Primitive.Int.read(data);
        let length = await Primitive.Int.read(data);
        let body = data.read(length);
        return new Message(await TLObject_js_1.TLObject.read(new BytesIO_js_1.BytesIO(body)), msgId, seqNo, length);
    }
    write() {
        let b = new BytesIO_js_1.BytesIO();
        b.write(toBytes(this.msgId));
        b.write(Primitive.Int.write(this.seqNo));
        b.write(Primitive.Int.write(this.length));
        b.write(this.body.write());
        return platform_node_js_1.Buffer.from(b.buffer);
    }
}
exports.Message = Message;
