// tgsnake/lib/src/TL/Messages/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.OrderInfo = exports.ShippingAddress = exports.PreCheckoutQuery = exports.ShippingQuery = exports.ChosenInlineResult = exports.InlineQuery = exports.Message = exports.ReplyMarkups = exports.Medias = void 0;
exports.Medias = __importStar(require("./Medias/index.js"));
exports.ReplyMarkups = __importStar(require("./ReplyMarkup.js"));
var Message_js_1 = require("./Message.js");
Object.defineProperty(exports, "Message", { enumerable: true, get: function () { return Message_js_1.Message; } });
var inlineQuery_js_1 = require("./inlineQuery.js");
Object.defineProperty(exports, "InlineQuery", { enumerable: true, get: function () { return inlineQuery_js_1.InlineQuery; } });
Object.defineProperty(exports, "ChosenInlineResult", { enumerable: true, get: function () { return inlineQuery_js_1.ChosenInlineResult; } });
var Shipping_js_1 = require("./Shipping.js");
Object.defineProperty(exports, "ShippingQuery", { enumerable: true, get: function () { return Shipping_js_1.ShippingQuery; } });
Object.defineProperty(exports, "PreCheckoutQuery", { enumerable: true, get: function () { return Shipping_js_1.PreCheckoutQuery; } });
Object.defineProperty(exports, "ShippingAddress", { enumerable: true, get: function () { return Shipping_js_1.ShippingAddress; } });
Object.defineProperty(exports, "OrderInfo", { enumerable: true, get: function () { return Shipping_js_1.OrderInfo; } });
